src/validationinterface.cpp

```diff
@@ -139,6 +139,11 @@ void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason
 }
 
 void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload) {
+    // Dependencies exist that require UpdatedBlockTip events to be delivered in the order in which
+    // the chain actually updates. In order to ensure this, these signals should be enqueued
+    // in the same critical section where the chain is updated
+    AssertLockHeld(cs_main);
```

Comment:
Bleh, if you're gonna assert this here you should probably assert it in pretty much every signal, as they all have ordering requirements. It also just seems kinda strange to do so cause its not really a requirement of the validationinterface, its a requirement that they be ordered, not done in one specific cs (eg the mempool ones could probably be ordered by mempool.cs instead).

Reply:
I was also not loving this AssertLockHeld() for the same reason.  I'd prefer we drop it and just leave a comment explaining the ordering requirement (maybe with a comment mentioning the cs_main lock as one way to achieve it).

Reply:
Agreed - updated

Reply:
I'm not a fan of spreading `AssertLockHeld` either but in this case the current fix is to lock and having the assert here makes that explicitly required. With just the comment a future refactor can break that requirement and reintroduce the bug.

But it was just a suggestion.

Reply:
Note that the `AssertLockHeld` wouldn't have prevented the caller from releasing cs_main after updating the tip, and reacquiring cs_main before the validation interface callbacks -- which would still result in violating the ordering requirements.

Reply:
@sdaftuar that's true and the same could be said to every `AssertLockHeld` 😛